spec/provider-spec.coffee

```diff
@@ -0,0 +1,283 @@
+            testProvider = new TestProvider(editor)
+            mainModule.registerProviderForEditor(testProvider, editor)
+            expect(autocompleteManager.scopes['source.js'][0]).toBe(testProvider)
+
+      describe "registerMultipleIdenticalProvidersForEditor", ->
+        it "registers the given provider once when called multiple times for the given editor", ->
+          runs ->
+            testProvider = new TestProvider(editor)
+            expect(autocompleteManager.scopes['source.js']).toBeUndefined()
+
+            mainModule.registerProviderForEditor(testProvider, editor)
+            expect(autocompleteManager.scopes['source.js'][0]).toBe(testProvider)
+            expect(_.size(autocompleteManager.scopes['source.js'])).toBe(1)
+
+            mainModule.registerProviderForEditor(testProvider, editor)
+            expect(autocompleteManager.scopes['source.js'][0]).toBe(testProvider)
+            expect(_.size(autocompleteManager.scopes['source.js'])).toBe(1)
+
+            mainModule.registerProviderForEditor(testProvider, editor)
+            expect(autocompleteManager.scopes['source.js'][0]).toBe(testProvider)
+            expect(_.size(autocompleteManager.scopes['source.js'])).toBe(1)
+
+      describe "unregisterProviderFromEditor", ->
+        it "unregisters the provider from all editors", ->
+          runs ->
+            testProvider = new TestProvider(editor)
+            mainModule.registerProviderForEditor(testProvider, editor)
+
+            expect(autocompleteManager.scopes['source.js'][0]).toBe(testProvider)
+
+            mainModule.unregisterProvider(testProvider)
+            expect(autocompleteManager.scopes).toEqual({})
+
+      describe "a registered provider", ->
+        it "calls buildSuggestions()", ->
+          runs ->
+            testProvider = new TestProvider(editor)
+            mainModule.registerProviderForEditor(testProvider, editor)
+
+            spyOn(testProvider, "buildSuggestions").andCallThrough()
+
+            # Trigger an autocompletion
+            editor.moveToBottom()
+            editor.moveToBeginningOfLine()
+            editor.insertText('f')
+            advanceClock completionDelay
+
+            expect(testProvider.buildSuggestions).toHaveBeenCalled()
+
+        it "calls confirm()", ->
+          runs ->
+            testProvider = new TestProvider(editor)
+            mainModule.registerProviderForEditor(testProvider, editor)
+
+            spyOn(testProvider, 'confirm').andCallThrough()
+
+            # Trigger an autocompletion
+            editor.moveToBottom()
+            editor.moveToBeginningOfLine()
+            editor.insertText('f')
+            advanceClock(completionDelay)
+
+            suggestionListView = atom.views.getView(autocompleteManager.suggestionList)
+            atom.commands.dispatch(suggestionListView, 'autocomplete-plus:confirm')
+
+            expect(testProvider.confirm).toHaveBeenCalled()
+
+    describe "Provider API v1.0.0", ->
```

Comment:
The existing provider API is in main.coffee, and 1.0.0 of the API is designed to match that. It will be deprecated completely, eventually, but this allows people to hop onto service-hub based registration without making drastic changes to their Provider in the short term.